### PR TITLE
Add 389 Directory Server image README

### PR DESCRIPTION
**Pull request by alexandrevicenzi**
This is related to #993.

### CODE REVIEW

File: src/bci_build/package/389-ds/README.md.j2 [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+By default, the image launches 389 Directory Server with the same configuration that comes with the SUSE Linux Enterprise Server, with one exception, a pre-configured Name Service Switch (NSS) configuration file (`/etc/nsswitch.conf`) in included.
+
+```ShellSession
+$ podman run -it --rm -p 3389:3389 -p 3636:3636 {{ image.pretty_reference }}
+```
+
+## Volumes
+
+The database is stored in the directory `/data`. A new empty database will be created on container startup unless an existing database is already present in `/data`.
```
Comment (dcermak) [inline src/bci_build/package/389-ds/README.md.j2:17]: Does this mean that `/data/` is a volume?

File: src/bci_build/package/389-ds/README.md.j2 [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+To mount a host directory as a volume for your database, run the following command:
+
+```ShellSession
+$ podman run -it --rm -v /my/own/datadir:/data:Z -p 3389:3389 -p 3636:3636 {{ image.pretty_reference }}
+```
+
+## Certificates
+
+By default, the container will use a self-signed CA certificate and a server certificate signed by that CA.
```
Comment (dcermak) [inline src/bci_build/package/389-ds/README.md.j2:27]: Will the server create its own CA?

File: src/bci_build/package/389-ds/README.md.j2 [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+A custom TLS certificate should be placed in `/data/tls/server.crt` and the key should be placed in and `/data/tls/server.key`.
```
Comment (dcermak) [inline src/bci_build/package/389-ds/README.md.j2:29]: We should mention in which format the certificate & key are expected to be (I guess it's PEM)

File: src/bci_build/package/389-ds/README.md.j2 [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+The CA certificates should be placed under `/data/tls/ca/` in separate files, i.e., `/data/tls/ca/ca1.crt` and `/data/tls/ca/ca2.crt`.
+
+## Environment variables
+
+### DS_ERRORLOG_LEVEL
+
+This optional environment variable can be used to set the log level for `ns-slapd` (default is `266354688`).
```
Comment (dcermak) [inline src/bci_build/package/389-ds/README.md.j2:36]: The log level is `266354688`? I think I need an explanation of their log level logic :sweat_smile:

File: src/bci_build/package/389-ds/README.md.j2 (hunk 1 above)
Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:17]: yes, `/data` is a volume. I guess this sentence is a bit vague right? It does not explicitly mention that it is a volume.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 2 above)
Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:27]: No, it does not, and neither has self-signed CAs.

This sentence is wrong for our image, it does have self-signed for official Quay images. I need to fix this, thanks for noticing.

```
INFO: Checking for PEM TLS files ...
INFO: Have /data/tls/server.key -> False
INFO: Have /data/tls/server.crt -> False
INFO: Have /data/tls/ca -> False
INFO: Have /data/config/pwdfile.txt -> True
INFO: Unable to configure TLS from PEM, missing a required file.
```

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:29]: Yes, good point, log messages mention PEM.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 1 above)
Comment (dcermak) [reply src/bci_build/package/389-ds/README.md.j2:17]: > I guess this sentence is a bit vague right? It does not explicitly mention that it is a volume.

It is not 100% clear that `/data/` is a volume. I'd rather mention it to be safe.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 4 above)
Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:36]: Oh, it seems this is the reason:

> levels are additive, to enable several different kinds of logging, add the values of those levels together. 

Take from https://access.redhat.com/documentation/en-us/red_hat_directory_server/11/html/configuration_command_and_file_reference/logs-reference

Not even sure what to document here, it seems no 389 DS docs include information about log levels, only OpenLDAP has it.

Comment (dcermak) [reply src/bci_build/package/389-ds/README.md.j2:36]: > Oh, it seems this is the reason:
> 
> > levels are additive, to enable several different kinds of logging, add the values of those levels together.
> 
> Take from https://access.redhat.com/documentation/en-us/red_hat_directory_server/11/html/configuration_command_and_file_reference/logs-reference
> 

But the linked page only mentions log levels up to 512. How do they end at these insanely large numbers?

If you can't find the info, then I'd rather leave it out to not confuse people even more.

Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:36]: A few more are documented here https://www.openldap.org/doc/admin24/slapdconfig.html.

The full level list is in the code https://github.com/389ds/389-ds-base/blob/main/ldap/servers/slapd/proto-slap.h#L809 but is not documented.

It is still worth documenting the value, but I'm not sure where I should send users to, 389ds seems to completely ignore documenting these values.

Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:36]: It does not hurt to ask https://github.com/389ds/389-ds-base/issues/6231.

Comment (dcermak) [reply src/bci_build/package/389-ds/README.md.j2:36]: Would you be fine with parking this for e.g. a week and see how upstream responds? If they don't then we just merge this without a proper explanation

Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:36]: I got feedback already, it covers most log levels. I can work around and document the levels on our side, using a markdown table (if supported in the registry).

I'm also fine waiting til next week to see if there are more comments on the upstream issue.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 1 above)
Comment (Firstyear) [reply src/bci_build/package/389-ds/README.md.j2:17]: Yes, /data is always a volume.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 2 above)
Comment (Firstyear) [reply src/bci_build/package/389-ds/README.md.j2:27]: Are you sure? The images should always create the self signed ca ?

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (Firstyear) [reply src/bci_build/package/389-ds/README.md.j2:29]: Yes, it's PEM.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 4 above)
Comment (Firstyear) [reply src/bci_build/package/389-ds/README.md.j2:36]: Seems like you already found all the log level docs, lmk if you need me to explain any others.

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:29]:
```suggestion
Place a custom TLS certificate in `/data/tls/server.crt` and the key in and `/data/tls/server.key`.
```

File: src/bci_build/package/389-ds/README.md.j2 [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+This optional environment variable can be used to run a database re-index task. Set the value to `1` to enable the task (default is disabled).
+
+### DS_SUFFIX_NAME
+
+This optional environment variable can be used to set the default database suffix name for `basedn` (default one is derived from the hostname).
+
+### DS_STARTUP_TIMEOUT
+
+This optional environment variable can be used to change the amount of time to wait for the instance to start (default is `60` seconds).
```
Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:57]:
```suggestion
Use this optional environment variable to change the time to wait for the instance to start (default is `60` seconds).
```

File: src/bci_build/package/389-ds/README.md.j2 [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+### DS_STOP_TIMEOUT
+
+This optional environment variable can be used to change the amount of time to wait for the instance to stop (default is `60` seconds).
```
Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:61]:
```suggestion
Use this optional environment variable to change the time to wait for the instance to stop (default is `60` seconds).
```

File: src/bci_build/package/389-ds/README.md.j2 [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+## Health, liveness, and readiness
+
+There is one explicit health check added to the container image. This check will verify if the service is misconfigured, if `ns-slapd` is running, and if the LDAPI is functional.
```
Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:65]:
```suggestion
The container image includes one explicit health check. This check will verify if the service is misconfigured, `ns-slapd` is running, and if the LDAPI is functional.
```

Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:65]: Which service?

File: src/bci_build/package/389-ds/README.md.j2 [hunk 8]
```diff
@@ -0,0 +1,67 @@
+# {{ image.pretty_name }} container image
+
+## Description
+
+[389 Directory Server](https://www.port389.org/) is a highly usable, fully featured, reliable and secure LDAP server implementation.
+
+## Usage
+
+By default, the image launches 389 Directory Server with the same configuration that comes with the SUSE Linux Enterprise Server, with one exception, a pre-configured Name Service Switch (NSS) configuration file (`/etc/nsswitch.conf`) in included.
```
Comment (lvicoun) [inline src/bci_build/package/389-ds/README.md.j2:9]:
```suggestion
By default, the image launches 389 Directory Server with the same configuration that comes with the SUSE Linux Enterprise Server, with one exception, a pre-configured Name Service Switch (NSS) configuration file (`/etc/nsswitch.conf`) is included.
```

**Review by lvicoun — APPROVED** [review]
Please consider the following suggestions. Thanks!

File: src/bci_build/package/389-ds/README.md.j2 (hunk 7 above)
Comment (alexandrevicenzi) [reply src/bci_build/package/389-ds/README.md.j2:65]: in this case is the 389-ds service.

File: src/bci_build/package/389-ds/README.md.j2 [hunk 9]
```diff
@@ -0,0 +1,67 @@
+# {{ image.pretty_name }} container image
+
+## Description
+
+[389 Directory Server](https://www.port389.org/) is a highly usable, fully featured, reliable and secure LDAP server implementation.
```
Comment (dmpop) [inline src/bci_build/package/389-ds/README.md.j2:5]: What does _highly usable_ mean? Can we drop it?

File: src/bci_build/package/389-ds/README.md.j2 (hunk 8 above)
Comment (dmpop) [inline src/bci_build/package/389-ds/README.md.j2:9]:
```suggestion
By default, the image launches 389 Directory Server with the same configuration that comes with the SUSE Linux Enterprise Server. However, there is also included a pre-configured Name Service Switch (NSS) configuration file (`/etc/nsswitch.conf`).
```

File: src/bci_build/package/389-ds/README.md.j2 (hunk 1 above)
Comment (dmpop) [inline src/bci_build/package/389-ds/README.md.j2:17]:
```suggestion
The database is stored in the directory `/data`. A new empty database is created during container startup, unless an existing database is already present in `/data`.
```

File: src/bci_build/package/389-ds/README.md.j2 (hunk 2 above)
Comment (dmpop) [inline src/bci_build/package/389-ds/README.md.j2:27]:
```suggestion
By default, the container uses a self-signed CA certificate and a server certificate signed by that CA.
```

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (dmpop) [inline src/bci_build/package/389-ds/README.md.j2:29]:
```suggestion
A custom TLS certificate must be placed in `/data/tls/server.crt`, and the key must be placed in `/data/tls/server.key`.
```

File: src/bci_build/package/389-ds/README.md.j2 (hunk 1 above)
Comment (dirkmueller) [reply src/bci_build/package/389-ds/README.md.j2]: added

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (dirkmueller) [reply src/bci_build/package/389-ds/README.md.j2]: added PEM

File: src/bci_build/package/389-ds/README.md.j2 (hunk 8 above)
Comment (dirkmueller) [reply src/bci_build/package/389-ds/README.md.j2]: merged manually

File: src/bci_build/package/389-ds/README.md.j2 (hunk 3 above)
Comment (dirkmueller) [reply src/bci_build/package/389-ds/README.md.j2]: changed to active voice.

**Review by dirkmueller — APPROVED** [review]

**Pull request merged**